# Lab | Intro to Probability

**Objective**

Welcome to this Intro to Probability lab, where we explore decision-making scenarios through the lens of probability and strategic analysis. In the business world, making informed decisions is crucial, especially when faced with uncertainties. This lab focuses on scenarios where probabilistic outcomes play a significant role in shaping strategies and outcomes. Students will engage in exercises that require assessing and choosing optimal paths based on data-driven insights. The goal is to enhance your skills by applying probability concepts to solve real-world problems.

**Challenge 1**

#### Ironhack Airlines 

Often Airlines sell more tickets than they have seats available, this is called overbooking. Consider the following:
- A plane has 450 seats. 
- Based on historical data we conclude that each individual passenger has a 3% chance of missing it's flight. 

If the Ironhack Airlines routinely sells 460 tickets, what is the chance that they have a seats for all passenger?

- Distribución binomial.
- n es el nº de billetes vendidos: 460
- p es la probabilidad de que el pasajero tenga su asiento.

In [1]:
from scipy.stats import binom

n = 460
p = 0.97

binom_dist = binom(n, p)

- Cálculo de la probabilidad habiendo más de 450 pasajeros.
- Cáculo de la probabilidad acumulada hasta 450 pasajeros, después hacemos 1 (100% - esa cantidad).

In [4]:
probabilidad_todos_tengan_asiento = 1 - binom_dist.cdf(450)
probabilidad_todos_tengan_asiento

0.11552275337845685

- 11.55% de probabilidad.

**Challenge 2**

#### Ironhack Call Center 

Suppose a customer service representative at a call center is handling customer complaints. Consider the following:
- The probability of successfully resolving a customer complaint on the first attempt is 0.3. 


What is the probability that the representative needs to make at least three attempts before successfully resolving a customer complaint?

- Distribución geométrica.
- Nº de intentos necesarios hasta el primer éxito en una serie de ensayos de Bernoulli, donde cada ensayo tiene una probabilidad de éxito constante.
- p = 0.3, es la probabilidad de resolver una queja en el primer intento.

- La probabilidad de que el primer intento falle es 1 − p = 0.7
- La probabilidad de que el segundo intento también falle es 0.7
- La probabilidad de que se necesiten al menos tres intentos es equivalente a la probabilidad de que los dos primeros intentos fallen.

- La probabilidad de necesitar al menos 3 intentos es: (1-p) x (1-p) = 0.7 x 0.7 = 0.49

In [5]:
from scipy.stats import geom

p = 0.3

probabilidad_al_menos_3_intentos = 1 - geom.cdf(2, p)
probabilidad_al_menos_3_intentos

0.49

- 49%

**Challenge 3**

#### Ironhack Website

Consider a scenario related to Ironhack website traffic. Where:
- our website takes on average 500 visits per hour.
- the website's server is designed to handle up to 550 vists per hour.


What is the probability of the website server being overwhelmed?

- Distribución Poisson.
- Conteo de eventos (visitas) que ocurren de manera independiente en un intervalo de tiempo fijo.

- Función de distribución acumulada (CDF) para calcular p(x < o igual a 550), luego restaremos ese valor de 1 para obtener p(x>550).

In [6]:
from scipy.stats import poisson

lambda_visitas = 500

probabilidad_sobrecarga = 1 - poisson.cdf(550, lambda_visitas)
probabilidad_sobrecarga

0.01289822084039205

- Primero calculamos la probabilidad acumulada de que el número de visitas por hora sea menor o igual a 550.
- Después restamos esta probabilidad de 1 para obtener la probabilidad de que el número de visitas exceda 550, lo que causaría una sobrecarga del servidor.

What is the probability of being overwhelmed at some point during a day? (consider 24hours)

- Probabilidad de no sobrecarga en una hora.

In [7]:
probabilidad_no_sobrecarga_una_hora = 1 - probabilidad_sobrecarga

- Probabilidad de no sobrecarga en 24 horas.

In [8]:
probabilidad_no_sobrecarga_dia = probabilidad_no_sobrecarga_una_hora ** 24

- Probabilidad de sobrecarga al menos una vez en el día

In [9]:
probabilidad_sobrecarga_dia = 1 - probabilidad_no_sobrecarga_dia
probabilidad_sobrecarga_dia

0.2677043869515715

- La probabilidad de que esto ocurra es el producto de las probabilidades de que el servidor no se sobrecargue en cada hora.

**Challenge 4**

#### Ironhack Helpdesk

Consider a scenario related to the time between arrivals of customers at a service desk.

On average, a customers arrives every 10minutes.

What is the probability that the next customer will arrive within the next 5 minutes?

- Distribución exponencial, tiempo entre eventos en un proceso de Poisson.
- El tiempo promedio entre la llegada de clientes es de 10 minutos, lo que nos da la tasa de llegada.
- La tasa de llegada es el inverso del tiempo promedio, es decir, 1 / 10 clientes por minuto.
- t es el tiempo en minutos.

In [10]:
import math

lambda_llegadas = 1 / 10
t = 5

probabilidad_llegada_5_minutos = 1 - math.exp(-lambda_llegadas * t)
probabilidad_llegada_5_minutos

0.3934693402873666

- lambda_llegadas es el inverso del tiempo promedio entre llegadas (10 minutos).
- Calculamos la probabilidad acumulada de que el próximo cliente llegue dentro de los próximos 5 minutos usando la fórmula de la distribución exponencial.
- Restamos la función exponencial de 1 para obtener la probabilidad acumulada.

If there is no customer for 15minutes, employees can that a 5minutes break.

What is the probability an employee taking a break?

- Para que un empleado pueda tomar un descanso de 5 minutos, no debe haber ninguna llegada de clientes en un intervalo de 15 minutos antes del descanso. 
- Necesitamos calcular la probabilidad de que no llegue ningún cliente en un período de 15 minutos.

In [11]:
t = 15

probabilidad_no_llegada_15_minutos = math.exp(-lambda_llegadas * t)
probabilidad_no_llegada_15_minutos

0.22313016014842982

- Probabilidad de que un empleado pueda tomar un descanso de 5 minutos, dado que no ha habido clientes durante los 15 minutos anteriores.

**Challenge 5**

The weights of a certain species of birds follow a normal distribution with a mean weight of 150 grams and a standard deviation of 10 grams. 

- If we randomly select a bird, what is the probability that its weight is between 140 and 160 grams?

- Distribución normal (Gaussiana).
- Distribución normal con media de 150 g.
- Desviación estándar de 10 g.
- Calculamos la probabilidad de que un pájarop al azar pese entre 140 y 160 g.

In [12]:
from scipy.stats import norm

mean = 150
std = 10

norm_dist = norm(loc=mean, scale=std)

prob_entre_140_y_160 = norm_dist.cdf(160) - norm_dist.cdf(140)

print(f"La probabilidad de peso entre 140 y 160 g es {prob_entre_140_y_160:.4f}")

La probabilidad de peso entre 140 y 160 g es 0.6827


- El resultado es la diferencia de las probabilidades acumuladas de que los pesos sean menores o iguales a 160 y 140 respectivamente.

**Challenge 6**

If the lifetime (in hours) of a certain electronic component follows an exponential distribution with a mean lifetime of 50 hours, what is the probability that the component fails within the first 30 hours?

- Distribución exponencial.
- Tiempo hasta el fallo de un componente cuando los fallos ocurren de manera continua.
- La tasa de fallo es el inverso del tiempo medio de vida: 1 / 50

In [13]:
mean_lifetime = 50
lambda_failure = 1 / mean_lifetime
time = 30 

prob_failure_within_30_hours = 1 - math.exp(-lambda_failure * time)
prob_failure_within_30_hours

0.4511883639059735

- math.exp(-lambda_failure * time) -> Calculamos la función exponencial que representa la probabilidad acumulada hasta t = 30 horas.
- Restamos a esa cantidad 1 para calcular la probabilidad de que el componente falle dentro de las primeras 30 horas.